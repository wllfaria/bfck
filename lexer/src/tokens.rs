#[derive(Debug, Clone, Copy)]
pub enum Tokens {
    MoveLeft(usize),
    MoveRight(usize),
    Increment(usize),
    Decrement(usize),
    Write(usize),
    Read(usize),
    JumpIfZero(usize),
    JumpUnlessZero(usize),
}

impl PartialEq for Tokens {
    fn eq(&self, other: &Self) -> bool {
        matches!(
            (self, other),
            (Tokens::MoveLeft(_), Tokens::MoveLeft(_))
                | (Tokens::MoveRight(_), Tokens::MoveRight(_))
                | (Tokens::Increment(_), Tokens::Increment(_))
                | (Tokens::Decrement(_), Tokens::Decrement(_))
                | (Tokens::Write(_), Tokens::Write(_))
                | (Tokens::Read(_), Tokens::Read(_))
                | (Tokens::JumpIfZero(_), Tokens::JumpIfZero(_))
                | (Tokens::JumpUnlessZero(_), Tokens::JumpUnlessZero(_))
        )
    }
}

impl Tokens {
    pub fn inner(&mut self) -> &mut usize {
        match self {
            Tokens::MoveLeft(count) => count,
            Tokens::MoveRight(count) => count,
            Tokens::Increment(count) => count,
            Tokens::Decrement(count) => count,
            Tokens::Write(count) => count,
            Tokens::Read(count) => count,
            Tokens::JumpIfZero(count) => count,
            Tokens::JumpUnlessZero(count) => count,
        }
    }
}

impl TryFrom<u8> for Tokens {
    type Error = &'static str;

    fn try_from(value: u8) -> Result<Tokens, Self::Error> {
        match value {
            b'<' => Ok(Tokens::MoveLeft(1)),
            b'>' => Ok(Tokens::MoveRight(1)),
            b'+' => Ok(Tokens::Increment(1)),
            b'-' => Ok(Tokens::Decrement(1)),
            b'.' => Ok(Tokens::Write(1)),
            b',' => Ok(Tokens::Read(1)),
            b'[' => Ok(Tokens::JumpIfZero(1)),
            b']' => Ok(Tokens::JumpUnlessZero(1)),
            _ => Err("not a valid token"),
        }
    }
}
