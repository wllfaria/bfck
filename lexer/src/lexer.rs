use crate::tokens::Tokens;

pub struct Lexer {}

impl Lexer {
    pub fn tokenize(input: &str) -> Vec<Tokens> {
        let input = input.as_bytes();
        let mut tokens = vec![];
        let mut index = 0;

        while index < input.len() {
            let token = Tokens::try_from(input[index]);

            // we couldn't convert the byte to a valid token, so we just consider it
            // a comment and skip over it
            if token.is_err() {
                index += 1;
                continue;
            }

            // its obviously safe to unwrap here since we check above, but lets make
            // it clear this is intended!
            assert!(token.is_ok(), "token was not valid yet we didn't skip it");
            let mut token = token.unwrap();

            // since we found a valid token, we will keep collecting tokens until we
            // find a different one
            index += 1;
            while index < input.len() {
                let next = Tokens::try_from(input[index]);

                // again, if the byte is a non-valid brainfuck token, we just skip
                // over it
                if next.is_err() {
                    index += 1;
                    continue;
                }

                // if it is a valid token, but its a different one from the one we
                // initially collected, we break out of this loop without
                // incrementing the index, so the next outer loop iteration will
                // resume from there
                //
                // its obviously safe to unwrap here since we check above, but lets make
                // it clear this is intended!
                assert!(next.is_ok(), "token was not valid yet we didn't skip it");
                if next.unwrap().ne(&token) {
                    break;
                }

                // if the tokens are the same, we increase the count of the token
                // through the inner() helper and increase the index to we do the
                // same with the next
                *token.inner() += 1;
                index += 1;
            }

            tokens.push(token);
        }

        tokens
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_input() {
        let tokens = Lexer::tokenize("     ");
        insta::assert_debug_snapshot!(tokens);
        assert!(tokens.is_empty());
    }

    #[test]
    fn single_spaced_token() {
        let tokens = Lexer::tokenize("<<       <<");
        insta::assert_debug_snapshot!(tokens);
        assert_eq!(tokens.last().unwrap(), &Tokens::MoveLeft(4));
    }

    #[test]
    fn changing_tokens() {
        let tokens = Lexer::tokenize("<<  >>");
        insta::assert_debug_snapshot!(tokens);
        assert_eq!(tokens.first().unwrap(), &Tokens::MoveLeft(2));
        assert_eq!(tokens.last().unwrap(), &Tokens::MoveRight(2));
    }

    #[test]
    fn ignoring_non_tokens() {
        let tokens = Lexer::tokenize(" _*  <<random_!?anything!!>!!>");
        insta::assert_debug_snapshot!(tokens);
        assert_eq!(tokens.first().unwrap(), &Tokens::MoveLeft(2));
    }

    #[test]
    fn pretty_hello_world() {
        let code = include_str!("../../samples/hello_world_pretty.bf");
        let tokens = Lexer::tokenize(code);
        insta::assert_debug_snapshot!(tokens);
        assert_eq!(tokens.len(), 58);
    }
}
