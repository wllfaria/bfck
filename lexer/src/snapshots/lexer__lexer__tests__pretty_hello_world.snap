---
source: lexer/src/lexer.rs
expression: tokens
---
[
    Increment(
        8,
    ),
    JumpIfZero(
        1,
    ),
    MoveRight(
        1,
    ),
    Increment(
        4,
    ),
    JumpIfZero(
        1,
    ),
    MoveRight(
        1,
    ),
    Increment(
        2,
    ),
    MoveRight(
        1,
    ),
    Increment(
        3,
    ),
    MoveRight(
        1,
    ),
    Increment(
        3,
    ),
    MoveRight(
        1,
    ),
    Increment(
        1,
    ),
    MoveLeft(
        4,
    ),
    Decrement(
        1,
    ),
    JumpUnlessZero(
        1,
    ),
    MoveRight(
        1,
    ),
    Increment(
        1,
    ),
    MoveRight(
        1,
    ),
    Increment(
        1,
    ),
    MoveRight(
        1,
    ),
    Decrement(
        1,
    ),
    MoveRight(
        2,
    ),
    Increment(
        1,
    ),
    JumpIfZero(
        1,
    ),
    MoveLeft(
        1,
    ),
    JumpUnlessZero(
        1,
    ),
    MoveLeft(
        1,
    ),
    Decrement(
        1,
    ),
    JumpUnlessZero(
        1,
    ),
    MoveRight(
        2,
    ),
    Write(
        1,
    ),
    MoveRight(
        1,
    ),
    Decrement(
        3,
    ),
    Write(
        1,
    ),
    Increment(
        7,
    ),
    Write(
        2,
    ),
    Increment(
        3,
    ),
    Write(
        1,
    ),
    MoveRight(
        2,
    ),
    Write(
        1,
    ),
    MoveLeft(
        1,
    ),
    Decrement(
        1,
    ),
    Write(
        1,
    ),
    MoveLeft(
        1,
    ),
    Write(
        1,
    ),
    Increment(
        3,
    ),
    Write(
        1,
    ),
    Decrement(
        6,
    ),
    Write(
        1,
    ),
    Decrement(
        8,
    ),
    Write(
        1,
    ),
    MoveRight(
        2,
    ),
    Increment(
        1,
    ),
    Write(
        1,
    ),
    MoveRight(
        1,
    ),
    Increment(
        2,
    ),
    Write(
        1,
    ),
]
